File: src/main/java/com/cout970/magneticraft/tileentity/TileCrushingTable.java
package com.cout970.magneticraft.tileentity;

import com.cout970.magneticraft.Magneticraft;
import com.cout970.magneticraft.api.access.RecipeCrushingTable;
import com.cout970.magneticraft.api.access.RecipeRegister;
import com.cout970.magneticraft.util.MiscUtils;
import net.darkaqua.blacksmith.Game;
import net.darkaqua.blacksmith.inventory.SimpleInventoryHandler;
import net.darkaqua.blacksmith.storage.IDataCompound;
import net.darkaqua.blacksmith.util.ResourceReference;
import net.darkaqua.blacksmith.vectors.Vect3d;
import net.minecraft.block.Block;
import net.minecraft.client.Minecraft;
import net.minecraft.client.audio.ISound;
import net.minecraft.client.audio.PositionedSoundRecord;
import net.minecraft.client.particle.EffectRenderer;
import net.minecraft.client.particle.EntityFX;
import net.minecraft.item.Item;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumParticleTypes;
import net.minecraftforge.items.IItemHandler;

/**
 * Created by cout970 on 16/12/2015.
 */
public class TileCrushingTable extends TileBase {

    private SimpleInventoryHandler inventory;
    private int progress;

    public TileCrushingTable() {
        inventory = new SimpleInventoryHandler(1) {

            @Override
            public void setStackInSlot(int slot, ItemStack stack) {
                super.setStackInSlot(slot, stack);
                sendUpdateToClient();
            }
        };
    }

    public IItemHandler getInventory() {
        return inventory;
    }

    @Override
    public void invalidate() {
        super.invalidate();
        if (Game.isServer()) {
            MiscUtils.dropItem(getWorldRef(), new Vect3d(0.5, 0.5, 0.5), inventory.getStackInSlot(0), true);
            inventory.setStackInSlot(0, null);
        }
    }

    @Override
    public void loadData(IDataCompound tag) {
        super.loadData(tag);
        inventory.load(tag, "inv");
    }

    @Override
    public void saveData(IDataCompound tag) {
        super.saveData(tag);
        inventory.save(tag, "inv");
    }

    public ItemStack getContent() {
        return inventory.getStackInSlot(0);
    }

    public void setContent(ItemStack input) {
        inventory.setStackInSlot(0, input);
    }

    public boolean canWork() {
        return (inventory.getStackInSlot(0) != null) &&
                (RecipeRegister.getCrushingTableRecipe(inventory.getStackInSlot(0)) != null);
    }

    public ItemStack getOutput() {
        RecipeCrushingTable rec = RecipeRegister.getCrushingTableRecipe(inventory.getStackInSlot(0));
        if (rec == null) { return null; }
        return rec.getOutput().copy();
    }

    public void tick(int maxHits) {
        progress++;
        if (Game.isClient()) {
            addParticles();
            if (progress != maxHits) { addHitSound(); }
        }
        if (progress >= maxHits) {
            progress = 0;
            setContent(getOutput().copy());
            addFinalSound();
        }
    }

    private void addHitSound() {
        ResourceReference res = new ResourceReference(Magneticraft.ID, "sounds.crushing_hit");
        Vect3d pos = new Vect3d(getPos()).add(0.5, 0.5, 0.5);
        ISound sound = PositionedSoundRecord.create(res.toResourceLocation(), (float) pos.getX(), (float) pos.getY(), (float) pos.getZ());
        Minecraft.getMinecraft().getSoundHandler().playSound(sound);
    }

    private void addFinalSound() {
        ResourceReference res = new ResourceReference(Magneticraft.ID, "sounds.crushing_final");
        Vect3d pos = new Vect3d(getPos()).add(0.5, 0.5, 0.5);
        ISound sound = PositionedSoundRecord.create(res.toResourceLocation(), (float) pos.getX(), (float) pos.getY(), (float) pos.getZ());
        Minecraft.getMinecraft().getSoundHandler().playSound(sound);
    }

    private void addParticles() {
        if (getContent() == null) { return; }
        EffectRenderer eff = Minecraft.getMinecraft().effectRenderer;
        Vect3d pos = new Vect3d(getPos()).add(0.5, 0.5, 0.5);
        for (int i = 0; i < 20; i++) {
            Vect3d motion = Vect3d.randomVector().multiply(1 / 32d);
            EntityFX particle;
            if (getContent().getItem() instanceof ItemBlock) {
                particle = eff.spawnEffectParticle(EnumParticleTypes.BLOCK_CRACK.getParticleID(), pos.getX(), pos.getY(), pos.getZ(),
                        motion.getX(), motion.getY(), motion.getZ(),  Block.getStateId(((ItemBlock) getContent().getItem()).block.getDefaultState()));
            } else {
                particle = eff.spawnEffectParticle(EnumParticleTypes.ITEM_CRACK.getParticleID(), pos.getX(), pos.getY(), pos.getZ(),
                        motion.getX(), motion.getY(), motion.getZ(), Item.getIdFromItem( getContent().getItem()),  getContent().getItemDamage());
            }
            particle.multipleParticleScaleBy(0.5f);
        }
    }
}

File: src/main/kotlin/com/cout970/magneticraft/tileentity/TileCrushingTable.kt
package com.cout970.magneticraft.tileentity

import com.cout970.magneticraft.api.internal.registries.machines.crushingtable.CrushingTableRecipeManager
import com.cout970.magneticraft.api.registries.machines.crushingtable.ICrushingTableRecipe
import com.cout970.magneticraft.misc.inventory.get
import com.cout970.magneticraft.misc.inventory.set
import com.cout970.magneticraft.misc.tileentity.shouldTick
import com.cout970.magneticraft.misc.world.isClient
import com.cout970.magneticraft.misc.world.isServer
import com.cout970.magneticraft.registry.ITEM_HANDLER
import com.cout970.magneticraft.registry.sounds
import com.cout970.magneticraft.util.add
import com.cout970.magneticraft.util.newNbt
import com.cout970.magneticraft.util.vector.*
import com.teamwizardry.librarianlib.common.util.autoregister.TileRegister
import net.minecraft.block.Block
import net.minecraft.client.Minecraft
import net.minecraft.client.particle.ParticleBreaking
import net.minecraft.client.particle.ParticleDigging
import net.minecraft.item.Item
import net.minecraft.item.ItemBlock
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumParticleTypes
import net.minecraft.util.ITickable
import net.minecraft.util.SoundCategory
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import net.minecraftforge.items.ItemStackHandler
import java.util.*

@TileRegister("crushing_table")
class TileCrushingTable : TileBase(), ITickable {

    val inventory: CrushingTableInventory = CrushingTableInventory()
    var damageTaken = 0

    private var recipeCache: ICrushingTableRecipe? = null
    private var inputCache: ItemStack? = null

    fun getRecipe(input: ItemStack): ICrushingTableRecipe? {
        if (input === inputCache) return recipeCache
        val recipe = CrushingTableRecipeManager.findRecipe(input)
        if (recipe != null) {
            recipeCache = recipe
            inputCache = input
        }
        return recipe
    }

    override fun update() {
        if (worldObj.isServer && shouldTick(100)) {
            sendUpdateToNearPlayers()
        }
    }

    fun getStack() = inventory[0]?.copy()

    fun setStack(stack: ItemStack?) {
        inventory[0] = stack?.copy()
    }

    fun canDamage() = getStack() != null && getRecipe(getStack()!!) != null

    fun doDamage(amount: Int) {
        if (!canDamage()) {
            return
        }

        damageTaken += amount

        if (damageTaken >= CRUSHING_DAMAGE) {
            if (world.isClient) {
                world.playSound(Minecraft.getMinecraft().thePlayer, pos, sounds["crushing_final"], SoundCategory.BLOCKS, 1F, 1F)
                spawnParticles()
            }

            inventory.setResult(getRecipe(getStack()!!)!!.output)
        } else if (world.isClient) {
            world.playSound(Minecraft.getMinecraft().thePlayer, pos, sounds["crushing_hit"], SoundCategory.BLOCKS, 1F, 1F)
            spawnParticles()
        }
        sendUpdateToNearPlayers()
    }

    @SideOnly(Side.CLIENT)
    private fun spawnParticles() {
        val center = pos.toVec3d() + vec3Of(0.5, 0.95, 0.5)
        val stack = getStack() ?: return
        val item = stack.item ?: return

        if (item is ItemBlock) {
            val state = item.block.getStateFromMeta(stack.metadata)
            val factory = ParticleDigging.Factory()
            val rand = Random()

            for (i in 0..5) {
                val particle = factory.createParticle(EnumParticleTypes.BLOCK_DUST.particleID, world, center.xd, center.yd, center.zd,
                        (rand.nextDouble() - 0.5) * 0.15, rand.nextDouble() * 0.2, (rand.nextDouble() - 0.5) * 0.15, Block.getStateId(state))
                Minecraft.getMinecraft().effectRenderer.addEffect(particle)
            }
        } else {
            val factory = ParticleBreaking.Factory()
            val rand = Random()

            for (i in 0..5) {
                val particle = factory.createParticle(EnumParticleTypes.BLOCK_DUST.particleID, world, center.xd, center.yd, center.zd,
                        (rand.nextDouble() - 0.5) * 0.15, rand.nextDouble() * 0.2, (rand.nextDouble() - 0.5) * 0.15, Item.getIdFromItem(item), stack.itemDamage)
                Minecraft.getMinecraft().effectRenderer.addEffect(particle)
            }
        }
    }

    override fun save(): NBTTagCompound {
        val nbt = newNbt {
            if (getStack() != null) {
                add("stack",getStack()!!.serializeNBT())
            }
            add("damage", damageTaken)
        }
        return super.save().also { it.merge(nbt) }
    }

    override fun load(nbt: NBTTagCompound) {
        inventory[0] = if (nbt.hasKey("stack")) {
            ItemStack.loadItemStackFromNBT(nbt.getCompoundTag("stack"))
        } else {
            null
        }
        damageTaken = nbt.getInteger("damage")
        super.load(nbt)
    }

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?) =
            (capability == ITEM_HANDLER) || super.hasCapability(capability, facing)

    @Suppress("UNCHECKED_CAST")
    override fun <T : Any> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
        return if (capability == ITEM_HANDLER) {
            inventory as T
        } else {
            super.getCapability(capability, facing)
        }
    }

    override fun onBreak() {
        super.onBreak()
        if (worldObj.isServer) {
            if (inventory[0] != null) {
                dropItem(inventory[0]!!, pos)
            }
        }
    }

    inner class CrushingTableInventory : ItemStackHandler(1) {
        override fun onContentsChanged(slot: Int) {
            damageTaken = 0
        }

        override fun getStackLimit(slot: Int, stack: ItemStack): Int = 1

        fun setResult(stack: ItemStack) {
            super.setStackInSlot(0, stack)
        }
    }

    companion object {
        val CRUSHING_DAMAGE = 40
    }
}